packages/injectable/ogre-tools-injectable.d.ts

```diff
@@ -1,150 +1,133 @@
+      Instance,
+      InstantiationParam
+    >(injectable: Injectable<InjectionInstance, Instance, InstantiationParam>): void;
     preventSideEffects: () => void;
   }
 
-  export interface DiContainerForSetup extends DiContainerForInjection<true> {}
+  export type Instantiate<Instance, InstantiationParam> = {
+    (di: DiContainerForInstantiate, param: InstantiationParam extends void
+      ? void
+      : InstantiationParam
+    ): Instance;
+  };
 
-  export interface DiContainerForInstantiate
-    extends DiContainerForInjection<false> {}
+  export type DiContainerForSetup = DiContainerForInjection<true>;
+  export type DiContainerForInstantiate = DiContainerForInjection<false>;
 
-  type InjectionToken<TInstance, TInstantiationParameter> = {
-    template: TInstance;
-    instantiationParameter: TInstantiationParameter;
+  export interface InjectionToken<
+    Instance,
+    InstantiationParam,
+  > {
+    template: Instance;
+    instantiationParameter: InstantiationParam;
     key: Symbol;
-  };
+  }
 
-  interface CommonInjectable {
+  export interface Injectable<
+    Instance extends InjectionInstance,
+    InjectionInstance,
+    InstantiationParam,
+  > {
     id: string;
     setup?: (di: DiContainerForSetup) => void | Promise<void>;
     causesSideEffects?: boolean;
-    lifecycle?: ILifecycle;
-  }
-
-  interface InjectionTokenInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
-  > extends CommonInjectable {
-    injectionToken: TInjectionToken;
-
-    instantiate: (
-      di: DiContainerForInstantiate,
-      instantiationParameter: InferFromToken<TInjectionToken>[1],
-    ) => InferFromToken<TInjectionToken>[0];
-  }
-
-  interface NormalInjectable<TInstance, TInstantiationParameter>
-    extends CommonInjectable {
-    instantiate: (
-      di: DiContainerForInstantiate,
-      instantiationParameter: TInstantiationParameter,
-    ) => TInstance;
+    injectionToken?: InjectionToken<InjectionInstance, InstantiationParam>;
+    instantiate: Instantiate<Instance, InstantiationParam>;
+    lifecycle: ILifecycle<
+      InstantiationParam,
+      string | number | symbol
+    >;
   }
 
   export function getInjectable<
-    TInjectionToken extends InjectionToken<unknown, unknown>,
+    Instance extends InjectionInstance,
+    InjectionInstance,
+    InstantiationParam = void
   >(
-    options: InjectionTokenInjectable<TInjectionToken>,
-  ): InjectionTokenInjectable<TInjectionToken>;
-
-  export function getInjectable<TInstance, TInstantiationParameter>(
-    options: NormalInjectable<TInstance, TInstantiationParameter>,
-  ): NormalInjectable<TInstance, TInstantiationParameter>;
-
-  export function getInjectionToken<TInstance, TInstantiationParameter = void>({
+    options: Omit<
+      Injectable<Instance, InjectionInstance, InstantiationParam>,
+      "lifecycle"
+    > & (
+      InstantiationParam extends void
+        ? {
+          lifecycle?: ILifecycle<void, string | number | symbol>
```

Comment:
Duplication for `string | number | symbol`. Or better yet, we could just not limit the type of an instance key, as there is no technical requirement for that.